app/file.go

```diff
@@ -623,32 +623,46 @@ func (a *App) UploadFileX(channelId, name string, input io.Reader,
 		}
 	}
 
-	aerr = t.readAll()
+	written, aerr := t.writeFile(io.MultiReader(t.buf, t.limitedInput), t.fileinfo.Path)
```

Comment:
Couple of questions here:

What is the objective here by writing before reading? Is it to just check the file upload size? If the file size is greater than the initial buffer size, how are we ensuring the file size limit then?

Sorry about so many questions, I'm a bit confused :)

 A comment would be helpful here.

Reply:
I think the code is a bit spread due to the `UploadFileTask` structure keeping most of the state. This can make it harder to follow. I have plans to improve this a bit in https://mattermost.atlassian.net/browse/MM-29081.

As far as your questions:

we write to the filestore to avoid loading the file in memory. The size limit should be guaranteed by the limited reader initialized in `t.init()`:

https://github.com/mattermost/mattermost-server/blob/155bd35e56a978a77f28f197878b7fdbcf1545da/app/file.go#L578-L581

Reply:
Ah okay, so instead of storing the file in memory, we are reading, writing, reading and writing again right?

I am wondering how much of an effect this has for S3 file stores. Would it be possible to do a comparison of time taken for small, medium and large files against localstore and S3?

Asking because I have a branch which does a streaming file read while serving emojis to leverage the `sendfile` system call. But I saw performance drops when using it for S3.

Reply:
> Ah okay, so instead of storing the file in memory, we are reading, writing, reading and writing again right?

Unless the plugin hook modifies the file we should be doing one write only to the backing store.
If the file is an image it's a bit different since we need to read it once more to allow for post-processing (preview and thumbnail generation). We might be able to optimize a bit such case a bit but I'd have to look deeper at how the image functions operate as far as memory allocations.

> I am wondering how much of an effect this has for S3 file stores. Would it be possible to do a comparison of time taken for small, medium and large files against localstore and S3?
> 
> Asking because I have a branch which does a streaming file read while serving emojis to leverage the `sendfile` system call. But I saw performance drops when using it for S3.

I can take a look but I imagine there will be some latency trade-off. After all there's a bit of back and forth especially with images and the plugin hook.

Reply:
> I can take a look but I imagine there will be some latency trade-off. After all there's a bit of back and forth especially with images and the plugin hook.

Yes, just want to see how much of a tradeoff is that. We definitely want to have this for sure. The test is more to get some data and confirm that the latency is what we expect it to be.

Reply:
So I might have gone a bit further than anticipated with this but I really wanted to have a clear picture.

It appears the `WriteFile()` call in `runPluginsHook()` is actually adding a significant amount of latency when S3 is used, even if no plugin is present so I'll see if I can find a way to trigger it from within the `RunMultiPluginHook()` function parameter. That should at least improve the default case. The image upload case has naturally worsened since we have to read data again from the filestore but that might be harder to improve.

![comparison](https://user-images.githubusercontent.com/1832946/94690805-d57d3c00-0330-11eb-8f31-a60e9334e05a.png)

Reply:
This looks great. Yeah improving the default case would be very helpful here.

Reply:
Much better now in case the plugin hook does not run:

![comparison_new](https://user-images.githubusercontent.com/1832946/94848660-9c25f880-0424-11eb-9387-cc2f5a758937.png)